executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,751 @@
+			task.doneErr = make(chan error, 1)
+			task.doneErr <- err
+		}
+		wg.Done()
+	}()
+
+	for ok := true; ok; {
+		select {
+		case task, ok = <-imw.taskCh:
+			if !ok {
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+
+		err := imw.handleTask(ctx, task)
+		task.doneErr <- err
+	}
+}
+
+func (imw *innerMergeWorker) handleTask(ctx context.Context, task *lookUpMergeJoinTask) error {
+	dLookUpKeys, err := imw.constructDatumLookupKeys(task)
+	if err != nil {
+		return err
+	}
+	dLookUpKeys = imw.dedupDatumLookUpKeys(dLookUpKeys)
+	if !imw.outerMergeCtx.needOuterSort && imw.desc {
+		lenKeys := len(dLookUpKeys)
+		for i := 0; i < lenKeys/2; i++ {
+			dLookUpKeys[i], dLookUpKeys[lenKeys-i-1] = dLookUpKeys[lenKeys-i-1], dLookUpKeys[i]
+		}
+	}
+	err = imw.fetchInnerResults(ctx, task, dLookUpKeys)
+	if err != nil {
+		return err
+	}
+
+	task.results = make(chan *chunk.Chunk, task.innerResult.NumChunks()+1)
+	task.wgSetResultsSize.Done()
+	err = imw.handleMergeJoin(ctx, task)
+	return err
+}
+
+func (imw *innerMergeWorker) handleMergeJoin(ctx context.Context, task *lookUpMergeJoinTask) (err error) {
+	chk := chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+	defer func() {
+		if chk.NumRows() > 0 {
+			select {
+			case task.results <- chk:
+			case <-ctx.Done():
+				return
+			}
+		}
+		return
+	}()
+	if task.innerResult.Len() == 0 {
+		for task.cursor < task.outerResult.NumRows() {
+			imw.joiner.onMissMatch(false, task.outerResult.GetRow(task.cursor), chk)
+			if chk.IsFull() {
+				select {
+				case task.results <- chk:
+				case <-ctx.Done():
+					return nil
+				}
+				chk = chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+			}
+			task.cursor++
+		}
+		return
+	}
+
+	task.innerIter = chunk.NewIterator4Chunk(task.innerResult.GetChunk(task.innerCursor))
+	task.innerIter.Begin()
+	initCmpResult := -1
+	if imw.innerMergeCtx.desc {
+		initCmpResult = 1
+	}
+	for task.cursor < task.outerResult.NumRows() {
+		outerRow := task.outerResult.GetRow(task.cursor)
+		if task.sameKeyIter == nil || task.sameKeyIter.Current() == task.sameKeyIter.End() {
+			cmpResult := initCmpResult
+			if len(task.outerMatch) > 0 && !task.outerMatch[task.cursor] {
+				task.cursor++
+				continue
+			}
+			if task.innerIter.Current() != task.innerIter.End() {
+				cmpResult, err = imw.compare(outerRow, task.innerIter.Current())
+				if err != nil {
+					return err
+				}
+			}
+			if (cmpResult >= 0 && !imw.innerMergeCtx.desc) || (cmpResult <= 0 && imw.innerMergeCtx.desc) {
+				err = imw.fetchNextOuterRows(task, outerRow)
+				if err != nil {
+					return
+				}
+			} else {
+				cmp := -1
+				if task.sameKeyIter != nil && len(task.sameKeyRows) > 0 {
+					cmp, err = imw.compare(outerRow, task.sameKeyIter.Begin())
+					if err != nil {
+						return
+					}
+				}
+				if cmp != 0 {
+					imw.joiner.onMissMatch(false, outerRow, chk)
+
+					task.cursor++
```

Comment:
What about the other outerRows with the same values, don't they need to be called onMissMatch?

Reply:
Of course, they need. I can't figure out why the iteration will miss "the other outerRows with the same values".